// @antv/chart-advisor/lib/ruler/rules/rules-for-graph/pred-node-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeTypeRules = exports.nodeTypePredRule = void 0;
var constants_1 = require("../../../constants");
var applyChartTypes = ['graph'];
exports.nodeTypePredRule = {
    id: 'pred-node-type',
    type: 'DESIGN',
    docs: {
        detailedText: '',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return applyChartTypes.indexOf(chartType) !== -1;
    },
    /**
     * @param dataProps
     * @returns TODO: Mapping function for node field -> node type e.g. if there is a combination of fields suitable for drawing pie charts then donut nodes should be recommended
     */
    optimizer: function (dataProps) {
        var type = constants_1.DEFAULT_NODE_TYPE;
        var graphInfo = dataProps.graphInfo, layoutType = dataProps.layoutType;
        if (graphInfo.nodeCount < 100 && layoutType === 'dagre') {
            type = 'rect';
        }
        return { nodeType: type };
    },
};
exports.nodeTypeRules = {
    'pred-node-type': exports.nodeTypePredRule,
};
