// @antv/chart-advisor/lib/ruler/rules/rules-for-graph/pred-edge-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.edgeTypeRules = exports.predEdgeTypeRule = void 0;
var applyChartTypes = ['graph'];
exports.predEdgeTypeRule = {
    id: 'pred-edge-type',
    type: 'DESIGN',
    docs: {
        detailedText: '',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return applyChartTypes.indexOf(chartType) !== -1;
    },
    optimizer: function (dataProps) {
        var _a = dataProps, graphInfo = _a.graphInfo, layoutType = _a.layoutType;
        var type = 'line';
        if ((graphInfo === null || graphInfo === void 0 ? void 0 : graphInfo.linkCount) < 60) {
            switch (layoutType) {
                case 'dagre': {
                    type = 'round';
                    break;
                }
                case 'radial': {
                    type = 'smooth';
                    break;
                }
                default: {
                    type = 'line';
                }
            }
        }
        return { edgeType: type };
    },
};
exports.edgeTypeRules = {
    'pred-edge-type': exports.predEdgeTypeRule,
};
