// @antv/chart-advisor/lib/ruler/rules/rules-for-graph/pred-scale-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scaleRules = exports.recScaleRule = void 0;
var data_wizard_1 = require("@antv/data-wizard");
var applyChartTypes = ['graph'];
exports.recScaleRule = {
    id: 'pred-scale-type',
    type: 'DESIGN',
    docs: {
        detailedText: '',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return applyChartTypes.indexOf(chartType) !== -1;
    },
    optimizer: function (fieldInfo) {
        var fieldEncodeType;
        if (!fieldInfo) {
            return 'linear';
        }
        if (data_wizard_1.analyzer.isOrdinal(fieldInfo) || data_wizard_1.analyzer.isNominal(fieldInfo)) {
            fieldEncodeType = 'ordinal';
        }
        else if (data_wizard_1.analyzer.isContinuous(fieldInfo) || fieldInfo.type === 'integer' || fieldInfo.type === 'float') {
            fieldEncodeType = 'linear';
        }
        return fieldEncodeType;
    },
};
exports.scaleRules = {
    'pred-scale-type': exports.recScaleRule,
};
