// @antv/antv-spec/lib/src/adaptor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.specToG6Config = exports.g6Render = exports.specToG6Plot = exports.specToG2PlotConfig = exports.g2plotRender = exports.specToG2Plot = void 0;
var g2plot_1 = require("./g2plot");
Object.defineProperty(exports, "specToG2Plot", { enumerable: true, get: function () { return g2plot_1.specToG2Plot; } });
Object.defineProperty(exports, "g2plotRender", { enumerable: true, get: function () { return g2plot_1.g2plotRender; } });
var toConfig_1 = require("./g2plot/toConfig");
Object.defineProperty(exports, "specToG2PlotConfig", { enumerable: true, get: function () { return toConfig_1.specToG2PlotConfig; } });
var g6_1 = require("./g6");
Object.defineProperty(exports, "specToG6Plot", { enumerable: true, get: function () { return g6_1.specToG6Plot; } });
Object.defineProperty(exports, "g6Render", { enumerable: true, get: function () { return g6_1.g6Render; } });
var toConfig_2 = require("./g6/toConfig");
Object.defineProperty(exports, "specToG6Config", { enumerable: true, get: function () { return toConfig_2.specToG6Config; } });
