// @antv/data-samples/lib/group/chartId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataByChartId = exports.DATA_SAMPLES_BY_CHART_ID = void 0;
var tslib_1 = require("tslib");
var data_1 = require("../data");
var name_1 = require("./name");
exports.DATA_SAMPLES_BY_CHART_ID = {
    // pie
    pie_chart: data_1.chart.pieChartTestData,
    donut_chart: data_1.chart.donutChartTestData,
    // line
    line_chart: data_1.chart.lineChartTestData,
    step_line_chart: data_1.chart.stepLineChartTestData,
    // area
    area_chart: data_1.chart.areaChartTestData,
    stacked_area_chart: data_1.chart.stackedAreaChartTestData,
    percent_stacked_area_chart: data_1.chart.stackedAreaChartTestData,
    // bar
    bar_chart: data_1.chart.barChartTestData,
    grouped_bar_chart: data_1.chart.groupedBarChartTestData,
    stacked_bar_chart: data_1.chart.stackedBarChartTestData,
    percent_stacked_bar_chart: data_1.chart.stackedBarChartTestData,
    // column
    column_chart: data_1.chart.columnChartTestData,
    grouped_column_chart: data_1.chart.groupedBarChartTestData,
    stacked_column_chart: data_1.chart.stackedBarChartTestData,
    percent_stacked_column_chart: data_1.chart.stackedBarChartTestData,
    // scatter
    scatter_plot: data_1.chart.scatterPlotTestData,
    // bubble
    bubble_chart: data_1.chart.bubbleChartTestData,
    // histogram
    histogram: data_1.chart.histogramTestData,
    // heatmap
    heatmap: data_1.chart.heatmapTestData,
};
function dataByChartId(chartId) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var dsName, data;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!exports.DATA_SAMPLES_BY_CHART_ID[chartId]) {
                        throw new Error("Data Sample for Chart ID ".concat(chartId, " is NOT available yet."));
                    }
                    dsName = exports.DATA_SAMPLES_BY_CHART_ID[chartId].name;
                    return [4 /*yield*/, (0, name_1.dataByName)(dsName)];
                case 1:
                    data = _a.sent();
                    return [2 /*return*/, data];
            }
        });
    });
}
exports.dataByChartId = dataByChartId;
