// @antv/ckb/lib/interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RECOMMEND_RATING_OPTIONS = exports.CHANNEL_OPTIONS = exports.LOM_OPTIONS = exports.SHAPE_OPTIONS = exports.GRAPHIC_CATEGORY_OPTIONS = exports.COORD_TYPE_OPTIONS = exports.PURPOSE_OPTIONS = exports.FAMILY_OPTIONS = void 0;
/**
 * @public
 */
exports.FAMILY_OPTIONS = [
    'LineCharts',
    'ColumnCharts',
    'BarCharts',
    'PieCharts',
    'AreaCharts',
    'ScatterCharts',
    'FunnelCharts',
    'HeatmapCharts',
    'RadarCharts',
    'TreeGraph',
    'GeneralGraph',
    'PolygonLayer',
    'LineLayer',
    'PointLayer',
    'HeatmapLayer',
    'Table',
    'Others',
];
/**
 * @public
 */
exports.PURPOSE_OPTIONS = [
    'Comparison',
    'Trend',
    'Distribution',
    'Rank',
    'Proportion',
    'Composition',
    'Relation',
    'Hierarchy',
    'Flow',
    'Spatial',
    'Anomaly',
    'Value',
];
/**
 * @public
 */
exports.COORD_TYPE_OPTIONS = [
    'NumberLine',
    'Cartesian2D',
    'SymmetricCartesian',
    'Cartesian3D',
    'Polar',
    'NodeLink',
    'Radar',
    'Geo',
    'Other',
];
/**
 * @public
 */
exports.GRAPHIC_CATEGORY_OPTIONS = ['Statistic', 'Diagram', 'Graph', 'Map', 'Other'];
/**
 * @public
 */
exports.SHAPE_OPTIONS = [
    'Lines',
    'Bars',
    'Round',
    'Square',
    'Area',
    'Scatter',
    'Symmetric',
    'Network',
    'Map',
    'Other',
];
/**
 * @public
 */
exports.LOM_OPTIONS = ['Nominal', 'Ordinal', 'Interval', 'Discrete', 'Continuous', 'Time'];
/**
 * @public
 */
exports.CHANNEL_OPTIONS = [
    'Position',
    'Length',
    'Color',
    'Area',
    'Angle',
    'ArcLength',
    'Direction',
    'Size',
    'Opacity',
    'Stroke',
    'LineWidth',
    'Lightness',
];
/**
 * @public
 */
exports.RECOMMEND_RATING_OPTIONS = ['Recommended', 'Use with Caution', 'Not Recommended'];
