// @antv/ckb/lib/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CKBOptions = void 0;
var tslib_1 = require("tslib");
var interface_1 = require("./interface");
var i18n_1 = require("./i18n");
/**
 * @param lang - Language of the options.
 * @public
 */
function CKBOptions(lang) {
    if (lang === void 0) { lang = 'en-US'; }
    if (lang && lang !== 'en-US') {
        var translator = i18n_1.I18N(lang);
        if (translator && translator.concepts) {
            var concepts = translator.concepts;
            return {
                family: Object.values(concepts.family),
                category: Object.values(concepts.category),
                purpose: Object.values(concepts.purpose),
                coord: Object.values(concepts.coord),
                shape: Object.values(concepts.shape),
                channel: Object.values(concepts.channel),
                lom: Object.values(concepts.lom),
                recRate: Object.values(concepts.recRate),
            };
        }
    }
    return {
        family: tslib_1.__spreadArray([], tslib_1.__read(interface_1.FAMILY_OPTIONS)),
        category: tslib_1.__spreadArray([], tslib_1.__read(interface_1.GRAPHIC_CATEGORY_OPTIONS)),
        purpose: tslib_1.__spreadArray([], tslib_1.__read(interface_1.PURPOSE_OPTIONS)),
        coord: tslib_1.__spreadArray([], tslib_1.__read(interface_1.COORD_TYPE_OPTIONS)),
        shape: tslib_1.__spreadArray([], tslib_1.__read(interface_1.SHAPE_OPTIONS)),
        channel: tslib_1.__spreadArray([], tslib_1.__read(interface_1.CHANNEL_OPTIONS)),
        lom: tslib_1.__spreadArray([], tslib_1.__read(interface_1.LOM_OPTIONS)),
        recRate: tslib_1.__spreadArray([], tslib_1.__read(interface_1.RECOMMEND_RATING_OPTIONS)),
    };
}
exports.CKBOptions = CKBOptions;
