// @antv/data-samples/esm/group/chartId.js
import { __awaiter, __generator } from "tslib";
import { chart as cd } from '../data';
import { dataByName } from './name';
export var DATA_SAMPLES_BY_CHART_ID = {
    // pie
    pie_chart: cd.pieChartTestData,
    donut_chart: cd.donutChartTestData,
    // line
    line_chart: cd.lineChartTestData,
    step_line_chart: cd.stepLineChartTestData,
    // area
    area_chart: cd.areaChartTestData,
    stacked_area_chart: cd.stackedAreaChartTestData,
    percent_stacked_area_chart: cd.stackedAreaChartTestData,
    // bar
    bar_chart: cd.barChartTestData,
    grouped_bar_chart: cd.groupedBarChartTestData,
    stacked_bar_chart: cd.stackedBarChartTestData,
    percent_stacked_bar_chart: cd.stackedBarChartTestData,
    // column
    column_chart: cd.columnChartTestData,
    grouped_column_chart: cd.groupedBarChartTestData,
    stacked_column_chart: cd.stackedBarChartTestData,
    percent_stacked_column_chart: cd.stackedBarChartTestData,
    // scatter
    scatter_plot: cd.scatterPlotTestData,
    // bubble
    bubble_chart: cd.bubbleChartTestData,
    // histogram
    histogram: cd.histogramTestData,
    // heatmap
    heatmap: cd.heatmapTestData,
};
export function dataByChartId(chartId) {
    return __awaiter(this, void 0, void 0, function () {
        var dsName, data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!DATA_SAMPLES_BY_CHART_ID[chartId]) {
                        throw new Error("Data Sample for Chart ID ".concat(chartId, " is NOT available yet."));
                    }
                    dsName = DATA_SAMPLES_BY_CHART_ID[chartId].name;
                    return [4 /*yield*/, dataByName(dsName)];
                case 1:
                    data = _a.sent();
                    return [2 /*return*/, data];
            }
        });
    });
}
