// @antv/thumbnails/lib/src/generated/thumbnails.js
"use strict";
// generated file thumbnails.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.STEP_LINE_CHART = exports.STACKED_COLUMN_CHART = exports.STACKED_BAR_CHART = exports.STACKED_AREA_CHART = exports.SCATTER_PLOT = exports.RADAR_CHART = exports.PIE_CHART = exports.PERCENT_STACKED_COLUMN_CHART = exports.PERCENT_STACKED_BAR_CHART = exports.PERCENT_STACKED_AREA_CHART = exports.LINE_CHART = exports.HISTOGRAM = exports.HEATMAP = exports.GROUPED_COLUMN_CHART = exports.GROUPED_BAR_CHART = exports.DONUT_CHART = exports.COLUMN_CHART = exports.BUBBLE_CHART = exports.BAR_CHART = exports.AREA_CHART = exports.isThumbnailID = exports.THUMBNAIL_IDS = void 0;
var tslib_1 = require("tslib");
var area_chart_1 = (0, tslib_1.__importDefault)(require("./charts/area_chart"));
exports.AREA_CHART = area_chart_1.default;
var bar_chart_1 = (0, tslib_1.__importDefault)(require("./charts/bar_chart"));
exports.BAR_CHART = bar_chart_1.default;
var bubble_chart_1 = (0, tslib_1.__importDefault)(require("./charts/bubble_chart"));
exports.BUBBLE_CHART = bubble_chart_1.default;
var column_chart_1 = (0, tslib_1.__importDefault)(require("./charts/column_chart"));
exports.COLUMN_CHART = column_chart_1.default;
var donut_chart_1 = (0, tslib_1.__importDefault)(require("./charts/donut_chart"));
exports.DONUT_CHART = donut_chart_1.default;
var grouped_bar_chart_1 = (0, tslib_1.__importDefault)(require("./charts/grouped_bar_chart"));
exports.GROUPED_BAR_CHART = grouped_bar_chart_1.default;
var grouped_column_chart_1 = (0, tslib_1.__importDefault)(require("./charts/grouped_column_chart"));
exports.GROUPED_COLUMN_CHART = grouped_column_chart_1.default;
var heatmap_1 = (0, tslib_1.__importDefault)(require("./charts/heatmap"));
exports.HEATMAP = heatmap_1.default;
var histogram_1 = (0, tslib_1.__importDefault)(require("./charts/histogram"));
exports.HISTOGRAM = histogram_1.default;
var line_chart_1 = (0, tslib_1.__importDefault)(require("./charts/line_chart"));
exports.LINE_CHART = line_chart_1.default;
var percent_stacked_area_chart_1 = (0, tslib_1.__importDefault)(require("./charts/percent_stacked_area_chart"));
exports.PERCENT_STACKED_AREA_CHART = percent_stacked_area_chart_1.default;
var percent_stacked_bar_chart_1 = (0, tslib_1.__importDefault)(require("./charts/percent_stacked_bar_chart"));
exports.PERCENT_STACKED_BAR_CHART = percent_stacked_bar_chart_1.default;
var percent_stacked_column_chart_1 = (0, tslib_1.__importDefault)(require("./charts/percent_stacked_column_chart"));
exports.PERCENT_STACKED_COLUMN_CHART = percent_stacked_column_chart_1.default;
var pie_chart_1 = (0, tslib_1.__importDefault)(require("./charts/pie_chart"));
exports.PIE_CHART = pie_chart_1.default;
var radar_chart_1 = (0, tslib_1.__importDefault)(require("./charts/radar_chart"));
exports.RADAR_CHART = radar_chart_1.default;
var scatter_plot_1 = (0, tslib_1.__importDefault)(require("./charts/scatter_plot"));
exports.SCATTER_PLOT = scatter_plot_1.default;
var stacked_area_chart_1 = (0, tslib_1.__importDefault)(require("./charts/stacked_area_chart"));
exports.STACKED_AREA_CHART = stacked_area_chart_1.default;
var stacked_bar_chart_1 = (0, tslib_1.__importDefault)(require("./charts/stacked_bar_chart"));
exports.STACKED_BAR_CHART = stacked_bar_chart_1.default;
var stacked_column_chart_1 = (0, tslib_1.__importDefault)(require("./charts/stacked_column_chart"));
exports.STACKED_COLUMN_CHART = stacked_column_chart_1.default;
var step_line_chart_1 = (0, tslib_1.__importDefault)(require("./charts/step_line_chart"));
exports.STEP_LINE_CHART = step_line_chart_1.default;
var Thumbnails = {
    area_chart: area_chart_1.default,
    bar_chart: bar_chart_1.default,
    bubble_chart: bubble_chart_1.default,
    column_chart: column_chart_1.default,
    donut_chart: donut_chart_1.default,
    grouped_bar_chart: grouped_bar_chart_1.default,
    grouped_column_chart: grouped_column_chart_1.default,
    heatmap: heatmap_1.default,
    histogram: histogram_1.default,
    line_chart: line_chart_1.default,
    percent_stacked_area_chart: percent_stacked_area_chart_1.default,
    percent_stacked_bar_chart: percent_stacked_bar_chart_1.default,
    percent_stacked_column_chart: percent_stacked_column_chart_1.default,
    pie_chart: pie_chart_1.default,
    radar_chart: radar_chart_1.default,
    scatter_plot: scatter_plot_1.default,
    stacked_area_chart: stacked_area_chart_1.default,
    stacked_bar_chart: stacked_bar_chart_1.default,
    stacked_column_chart: stacked_column_chart_1.default,
    step_line_chart: step_line_chart_1.default,
};
exports.default = Thumbnails;
exports.THUMBNAIL_IDS = [
    'area_chart',
    'bar_chart',
    'bubble_chart',
    'column_chart',
    'donut_chart',
    'grouped_bar_chart',
    'grouped_column_chart',
    'heatmap',
    'histogram',
    'line_chart',
    'percent_stacked_area_chart',
    'percent_stacked_bar_chart',
    'percent_stacked_column_chart',
    'pie_chart',
    'radar_chart',
    'scatter_plot',
    'stacked_area_chart',
    'stacked_bar_chart',
    'stacked_column_chart',
    'step_line_chart',
];
function isThumbnailID(id) {
    return exports.THUMBNAIL_IDS.includes(id);
}
exports.isThumbnailID = isThumbnailID;
