// @antv/antv-spec/lib/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.g6Render = exports.g2plotRender = exports.specToG6Config = exports.specToG2PlotConfig = exports.specToG6Plot = exports.specToG2Plot = exports.version = void 0;
const tslib_1 = require("tslib");
const package_json_1 = (0, tslib_1.__importDefault)(require("../package.json"));
exports.version = package_json_1.default.version;
(0, tslib_1.__exportStar)(require("./schema"), exports);
var adaptor_1 = require("./adaptor");
Object.defineProperty(exports, "specToG2Plot", { enumerable: true, get: function () { return adaptor_1.specToG2Plot; } });
Object.defineProperty(exports, "specToG6Plot", { enumerable: true, get: function () { return adaptor_1.specToG6Plot; } });
Object.defineProperty(exports, "specToG2PlotConfig", { enumerable: true, get: function () { return adaptor_1.specToG2PlotConfig; } });
Object.defineProperty(exports, "specToG6Config", { enumerable: true, get: function () { return adaptor_1.specToG6Config; } });
Object.defineProperty(exports, "g2plotRender", { enumerable: true, get: function () { return adaptor_1.g2plotRender; } });
Object.defineProperty(exports, "g6Render", { enumerable: true, get: function () { return adaptor_1.g6Render; } });
