// @antv/chart-advisor/lib/constants/graph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_LAYOUT_TYPE = exports.DEFAULT_NODE_TYPE = exports.DEGREE_FOR_SPARSE_GRAPH = exports.MAX_NUM_NODES_A_LEVEL = exports.DEFAULT_EDGE_WIDTH_RANGE = exports.DEFAULT_NODE_SIZE_RANGE = exports.MAX_DISTINCT_CLUSTER = exports.MAX_DISTINCT_COLOR = exports.ALL_LAYOUT_TYPES = exports.DEFAULT_COLORS = void 0;
exports.DEFAULT_COLORS = [
    '#BDD2FD',
    '#BDEFDB',
    '#C2C8D5',
    '#FBE5A2',
    '#F6C3B7',
    '#B6E3F5',
    '#D3C6EA',
    '#FFD8B8',
    '#AAD8D8',
    '#FFD6E7',
];
exports.ALL_LAYOUT_TYPES = ['force', 'grid', 'dagre', 'circular', 'concentric', 'radial'];
exports.MAX_DISTINCT_COLOR = exports.DEFAULT_COLORS.length;
exports.MAX_DISTINCT_CLUSTER = exports.DEFAULT_COLORS.length;
exports.DEFAULT_NODE_SIZE_RANGE = [15, 30];
exports.DEFAULT_EDGE_WIDTH_RANGE = [1, 10];
exports.MAX_NUM_NODES_A_LEVEL = 16;
exports.DEGREE_FOR_SPARSE_GRAPH = 3;
exports.DEFAULT_NODE_TYPE = 'point';
exports.DEFAULT_LAYOUT_TYPE = 'force';
