// @antv/chart-advisor/esm/chart-advisor/index.js
import { __assign, __read, __spreadArray } from "tslib";
import { Advisor } from '../advisor';
import { Linter } from '../linter';
var ChartAdvisor = /** @class */ (function () {
    function ChartAdvisor(config) {
        if (config === void 0) { config = {}; }
        this.advisor = new Advisor(config);
        this.linter = new Linter(config.ruleCfg);
    }
    ChartAdvisor.prototype.advise = function (params) {
        var result = this.adviseWithLog(params);
        var advices = result.advices;
        return advices;
    };
    /**
     * Advising charts by data and providing linting results for each chart
     */
    ChartAdvisor.prototype.adviseWithLog = function (params) {
        var _this = this;
        var dataProps = params.dataProps, options = params.options;
        // advising
        var adviceResult = this.advisor.adviseWithLog(params);
        var advices = adviceResult.advices, log = adviceResult.log;
        // linting
        var advicesAfterLint = advices.map(function (advice) {
            var _a;
            // No lint suggestions for graph visualization for now
            if (advice.type === 'graph') {
                return __assign(__assign({}, advice), { lint: [] });
            }
            var _b = _this.linter.lintWithLog({
                spec: advice.spec,
                dataProps: dataProps,
                options: options,
            }), lints = _b.lints, lintLog = _b.log;
            (_a = log.find(function (l) { return l.chartType === advice.type; }).log).push.apply(_a, __spreadArray([], __read(lintLog)));
            return __assign(__assign({}, advice), { lint: lints });
        });
        var result = {
            advices: advicesAfterLint,
            log: log,
        };
        return result;
    };
    return ChartAdvisor;
}());
export { ChartAdvisor };
