// @antv/chart-advisor/lib/ruler/rules/bar-without-axis-min.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.barWithoutAxisMin = void 0;
var applyChartTypes = [
    'bar_chart',
    'grouped_bar_chart',
    'stacked_bar_chart',
    'percent_stacked_bar_chart',
    'column_chart',
    'grouped_column_chart',
    'stacked_column_chart',
    'percent_stacked_column_chart',
];
exports.barWithoutAxisMin = {
    id: 'bar-without-axis-min',
    type: 'DESIGN',
    docs: {
        lintText: 'It is not recommended to set the minimum value of axis for the bar or column chart.',
        fixText: 'Remove the minimum value config of axis.',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return applyChartTypes.includes(chartType);
    },
    optimizer: function (_, chartSpec) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        var layer = chartSpec.layer;
        var xMin = (_d = (_c = (_b = (_a = layer === null || layer === void 0 ? void 0 : layer[0]) === null || _a === void 0 ? void 0 : _a.encoding) === null || _b === void 0 ? void 0 : _b.x) === null || _c === void 0 ? void 0 : _c.axis) === null || _d === void 0 ? void 0 : _d.min;
        var yMin = (_h = (_g = (_f = (_e = layer === null || layer === void 0 ? void 0 : layer[0]) === null || _e === void 0 ? void 0 : _e.encoding) === null || _f === void 0 ? void 0 : _f.y) === null || _g === void 0 ? void 0 : _g.axis) === null || _h === void 0 ? void 0 : _h.min;
        if (xMin || yMin) {
            var fixedLayer = JSON.parse(JSON.stringify(layer));
            if (xMin)
                fixedLayer[0].encoding.x.axis.min = 0;
            if (yMin)
                fixedLayer[0].encoding.y.axis.min = 0;
            return { layer: fixedLayer };
        }
        return {};
    },
};
