// @antv/chart-advisor/lib/ruler/rules/data-check.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataCheck = void 0;
var ckb_1 = require("@antv/ckb");
var utils_1 = require("../utils");
var Wiki = ckb_1.CKBJson('en-US', true);
var allChartTypes = Object.keys(Wiki);
exports.dataCheck = {
    id: 'data-check',
    type: 'HARD',
    docs: {
        lintText: 'Data must satisfy the data prerequisites.',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return allChartTypes.includes(chartType);
    },
    validator: function (args) {
        var result = 0;
        var dataProps = args.dataProps, chartType = args.chartType;
        if (dataProps && chartType && Wiki[chartType]) {
            result = 1;
            var dataPres = Wiki[chartType].dataPres || [];
            dataPres.forEach(function (dataPre) {
                if (!utils_1.verifyDataProps(dataPre, dataProps)) {
                    result = 0;
                }
                return true;
            });
        }
        return result;
    },
};
