// @antv/chart-advisor/lib/ruler/rules/diff-pie-sector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.diffPieSector = void 0;
var utils_1 = require("../../utils");
var applyChartTypes = ['pie_chart', 'donut_chart'];
exports.diffPieSector = {
    id: 'diff-pie-sector',
    type: 'SOFT',
    docs: {
        lintText: 'The difference between sectors of a pie chart should be large enough.',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return applyChartTypes.includes(chartType);
    },
    validator: function (args) {
        var result = 0;
        var dataProps = args.dataProps;
        if (dataProps) {
            var intervalField = dataProps.find(function (field) { return utils_1.hasSubset(field.levelOfMeasurements, ['Interval']); });
            if (intervalField && intervalField.sum && intervalField.rawData) {
                var sum = intervalField.sum;
                var scale_1 = 1 / sum;
                var scaledSamples = intervalField.rawData.map(function (v) { return v * scale_1; });
                var scaledProduct = scaledSamples.reduce(function (a, c) { return a * c; });
                var count = intervalField.rawData.length;
                var maxProduct = Math.pow((1 / count), count);
                result = Math.abs(maxProduct - Math.abs(scaledProduct)) / maxProduct;
            }
        }
        return result;
    },
};
