// @antv/chart-advisor/lib/ruler/rules/no-redundant-field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noRedundantField = void 0;
var ckb_1 = require("@antv/ckb");
var Wiki = ckb_1.CKBJson('en-US', true);
var allChartTypes = Object.keys(Wiki);
exports.noRedundantField = {
    id: 'no-redundant-field',
    type: 'HARD',
    docs: {
        lintText: 'No redundant field.',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return allChartTypes.includes(chartType);
    },
    validator: function (args) {
        var result = 0;
        var dataProps = args.dataProps, chartType = args.chartType;
        if (dataProps && chartType && Wiki[chartType]) {
            var dataPres = Wiki[chartType].dataPres || [];
            var maxFieldQty = dataPres
                .map(function (e) {
                if (e.maxQty === '*') {
                    return 99;
                }
                return e.maxQty;
            })
                .reduce(function (acc, cv) { return acc + cv; });
            if (dataProps.length) {
                var fieldQty = dataProps.length;
                if (fieldQty <= maxFieldQty) {
                    result = 1;
                }
            }
        }
        return result;
    },
};
