// @antv/chart-advisor/lib/ruler/rules/x-axis-line-fading.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xAxisLineFading = void 0;
var applyChartTypes = ['line_chart'];
exports.xAxisLineFading = {
    id: 'x-axis-line-fading',
    type: 'DESIGN',
    docs: {
        // FIXME formal description
        lintText: 'Adjust axis to make it prettier',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return applyChartTypes.includes(chartType);
    },
    optimizer: function (dataProps, chartSpec) {
        var _a;
        var layerEnc = chartSpec.layer && 'encoding' in chartSpec.layer[0] ? chartSpec.layer[0].encoding : null;
        if (layerEnc && ((_a = layerEnc.y) === null || _a === void 0 ? void 0 : _a.type) === 'quantitative') {
            var fieldInfo = dataProps.find(function (item) { var _a; return item.name === ((_a = layerEnc.y) === null || _a === void 0 ? void 0 : _a.field); });
            if (fieldInfo) {
                var range = fieldInfo.maximum - fieldInfo.minimum;
                if (fieldInfo.minimum && fieldInfo.maximum && range < (fieldInfo.maximum * 2) / 3) {
                    var yScaleMin = Math.floor(fieldInfo.minimum - range / 5);
                    return {
                        encoding: {
                            x: {
                                axis: {
                                    ticks: false,
                                    domain: false,
                                },
                            },
                            y: {
                                scale: {
                                    domainMin: yScaleMin > 0 ? yScaleMin : 0,
                                },
                            },
                        },
                    };
                }
            }
        }
        return {};
    },
};
