// @antv/chart-advisor/esm/ruler/rules/bar-series-qty.js
import { hasSubset } from '../../utils';
var applyChartTypes = [
    'bar_chart',
    'grouped_bar_chart',
    'stacked_bar_chart',
    'percent_stacked_bar_chart',
    'column_chart',
    'grouped_column_chart',
    'stacked_column_chart',
    'percent_stacked_column_chart',
];
export var barSeriesQty = {
    id: 'bar-series-qty',
    type: 'SOFT',
    docs: {
        lintText: 'Bar chart should has proper number of bars or bar groups.',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return applyChartTypes.includes(chartType);
    },
    validator: function (args) {
        var result = 0;
        var dataProps = args.dataProps, chartType = args.chartType;
        if (dataProps && chartType) {
            var field4Series = dataProps.find(function (field) { return hasSubset(field.levelOfMeasurements, ['Nominal']); });
            var seriesQty = field4Series && field4Series.count ? field4Series.count : 0;
            // TODO limit for this rule
            if (seriesQty >= 2 && seriesQty <= 20) {
                result = 1;
            }
            else if (seriesQty > 20) {
                result = 20 / seriesQty;
            }
        }
        return result;
    },
};
