// @antv/chart-advisor/esm/ruler/rules/data-field-qty.js
import { CKBJson } from '@antv/ckb';
var Wiki = CKBJson('en-US', true);
var allChartTypes = Object.keys(Wiki);
export var dataFieldQty = {
    id: 'data-field-qty',
    type: 'HARD',
    docs: {
        lintText: 'Data must have at least the min qty of the prerequisite.',
    },
    trigger: function (_a) {
        var chartType = _a.chartType;
        return allChartTypes.indexOf(chartType) !== -1;
    },
    validator: function (args) {
        var result = 0;
        var dataProps = args.dataProps, chartType = args.chartType;
        if (dataProps && chartType && Wiki[chartType]) {
            result = 1;
            var dataPres = Wiki[chartType].dataPres || [];
            var minFieldQty = dataPres.map(function (e) { return e.minQty; }).reduce(function (acc, cv) { return acc + cv; });
            if (dataProps.length) {
                var fieldQty = dataProps.length;
                if (fieldQty >= minFieldQty) {
                    result = 1;
                }
            }
        }
        return result;
    },
};
